### PR TITLE
Faster schematic discover from jar files

### PR DESCRIPTION
**Pull request by Nightenom**
# Changes proposed in this pull request:
- reworked loading from jars based on [patch](https://gist.github.com/malte0811/feec55ca151f13f99618654a2b258a09) by @malte0811
- removed unnecessary duplicated getStream() calls

Review please

### CODE REVIEW

File: src/main/java/com/ldtteam/structurize/management/Structures.java [hunk 1]
```diff
@@ -170,95 +158,6 @@ private static void loadStyleMaps()
...
-                final URL url = Structures.class.getResource(SCHEMATICS_ASSET_PATH + origin);
-                uri = url.toURI();
-            }
-            catch (@NotNull final URISyntaxException e)
-            {
-                Log.getLogger().error("loadStyleMaps : ", e);
-                return;
-            }
-
-            if ("jar".equals(uri.getScheme()))
```
Comment (Raycoms) [inline src/main/java/com/ldtteam/structurize/management/Structures.java:193]: Don't we need this anymore? I know that a lot of this was compatibility for different OS

Comment (Raycoms) [reply src/main/java/com/ldtteam/structurize/management/Structures.java:193]: + Different origin for dev env and for real game

**Comment by Nightenom** [comment]
@malte0811 your review would be appreciated

File: src/main/java/com/ldtteam/structurize/util/StructureLoadingUtils.java [hunk 2]
```diff
@@ -28,7 +32,16 @@
     /**
      * The list of origin folders.
      */
-    public static List<String> originFolders = new ArrayList<>();
+    private static final List<String> originFolders = new ArrayList<>();
+    /**
+     * The list of origin mods.
+     */
+    private static final Map<String, ModFileInfo> originMods = new HashMap<>();
```
Comment (malte0811) [inline src/main/java/com/ldtteam/structurize/util/StructureLoadingUtils.java:39]: Theoretically you can get something like mod A adding resources in mod B's namespace (so you'd need to check all mods to know a resource doesn't exist), but I guess you are intentionally ignoring that possibility?

**Review by malte0811 — DISMISSED** [review]
Looks good to me.

File: src/main/java/com/ldtteam/structurize/util/StructureLoadingUtils.java (hunk 2 above)
Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/util/StructureLoadingUtils.java:39]: Yep, I wondered why you were putting the result from searching through all mods into the modid<->modfile map. You can now completely override mod entry (in this map) by addOriginMod() method. I would like to use resource packs instead (and remove this mess) but I was told that there is some problem with that...

File: src/main/java/com/ldtteam/structurize/management/Structures.java (hunk 1 above)
Comment (marchermans) [reply src/main/java/com/ldtteam/structurize/management/Structures.java:193]: This is still needed if he does not handle it differently elsewhere.

File: src/main/java/com/ldtteam/structurize/management/Structures.java [hunk 3]
```diff
@@ -297,11 +198,12 @@ private static void checkDirectory(@NotNull final File directory)
...
      * and add them in the md5Map
      *
-     * @param basePath the base path.
+     * @param base     the base path.
      * @param prefix   either schematics, scans, cache
      */
-    private static void loadSchematicsForPrefix(@NotNull final Path basePath, @NotNull final String prefix)
+    private static void loadSchematicsForPrefix(@NotNull final Path base, @NotNull final String prefix)
     {
+        final Path basePath = base.toAbsolutePath();
```
Comment (marchermans) [inline src/main/java/com/ldtteam/structurize/management/Structures.java:206]: Not going to work in an obfuscated environment.
This needs to take the scheme into account -> Which means you need to use at least a URI.

**Review by marchermans — CHANGES_REQUESTED** [review]
Personally: This is not the way to go!
If this is reworked, this should be using datapacks and IFutureReloadListeners to load everything asynchronously and then process then on the main thread if even needed (if you implement the backing registry with concurrency in mind then you can do it all asynchronously).

That said, the new code will not work once packed into a jar. Since Paths do not care for the fact that it is in a jar or not.

File: src/main/java/com/ldtteam/structurize/management/Structures.java (hunk 3 above)
Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/management/Structures.java:206]: It works, it's tested well

**Comment by Nightenom** [comment]
Orion: load datapack from dedi server (to client) or load resource pack on dedi server, if you can manage to get one this working then I would like to use it too. Since either of these seems to be not feasible to me
btw this is not a rework, it's just two fixes: one on our end and the second is to mitigate forge behaviour by handling it completely on our end

File: src/main/java/com/ldtteam/structurize/management/Structures.java (hunk 3 above)
Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/management/Structures.java:206]: Btw exactly this was added to fix handling of the damned ZipPath (which is used when loading from jar) which was not working well with relativize

**Comment by Nightenom** [comment]
> That said, the new code will not work once packed into a jar. Since Paths do not care for the fact that it is in a jar or not.

Nope, wrong, afaik cpw added handling for this with the new mod launcher. If you look carefully this fix is using modFile locator which is also used for mods.toml/accesstransformer.cfg discover and few others

**Comment by marchermans** [comment]
Oh okey
If that is added then I am fine,
did not know that.

Learned something today.

**Pull request merged**